core/src/main/scala/akka/persistence/r2dbc/internal/h2/H2JournalDao.scala

```diff
@@ -0,0 +1,364 @@
+
+    if (log.isDebugEnabled)
+      result.foreach(seqNr => log.debug("Lowest sequence nr for persistenceId [{}]: [{}]", persistenceId, seqNr))
+
+    result
+  }
+
+  private def highestSeqNrForDelete(persistenceId: String, toSequenceNr: Long): Future[Long] = {
+    if (toSequenceNr == Long.MaxValue)
+      readHighestSequenceNr(persistenceId, 0L)
+    else
+      Future.successful(toSequenceNr)
+  }
+
+  private def lowestSequenceNrForDelete(persistenceId: String, toSeqNr: Long, batchSize: Int): Future[Long] = {
+    if (toSeqNr <= batchSize) {
+      Future.successful(1L)
+    } else {
+      readLowestSequenceNr(persistenceId)
+    }
+  }
+
+  def deleteEventsTo(persistenceId: String, toSequenceNr: Long, resetSequenceNumber: Boolean): Future[Unit] = {
+
+    def insertDeleteMarkerStmt(deleteMarkerSeqNr: Long, connection: Connection): Statement = {
+      val entityType = PersistenceId.extractEntityType(persistenceId)
+      val slice = persistenceExt.sliceForPersistenceId(persistenceId)
+      connection
+        .createStatement(insertDeleteMarkerSql)
+        .bind(0, slice)
+        .bind(1, entityType)
+        .bind(2, persistenceId)
+        .bind(3, deleteMarkerSeqNr)
+        .bind(4, "")
+        .bind(5, "")
+        .bind(6, 0)
+        .bind(7, "")
+        .bindPayloadOption(8, None)
+        .bind(9, true)
+    }
+
+    def deleteBatch(from: Long, to: Long, lastBatch: Boolean): Future[Unit] = {
+      (if (lastBatch && !resetSequenceNumber) {
+         r2dbcExecutor
+           .update(s"delete [$persistenceId] and insert marker") { connection =>
+             Vector(
+               connection.createStatement(deleteEventsSql).bind(0, persistenceId).bind(1, from).bind(2, to),
+               insertDeleteMarkerStmt(to, connection))
+           }
+           .map(_.head)
+       } else {
+         r2dbcExecutor
+           .updateOne(s"delete [$persistenceId]") { connection =>
+             connection.createStatement(deleteEventsSql).bind(0, persistenceId).bind(1, from).bind(2, to)
+           }
+       }).map(deletedRows =>
+        if (log.isDebugEnabled) {
+          log.debugN(
+            "Deleted [{}] events for persistenceId [{}], from seq num [{}] to [{}]",
+            deletedRows,
+            persistenceId,
+            from,
+            to)
+        })(ExecutionContexts.parasitic)
+    }
+
+    val batchSize = journalSettings.cleanupSettings.eventsJournalDeleteBatchSize
+
+    def deleteInBatches(from: Long, maxTo: Long): Future[Unit] = {
+      if (from + batchSize > maxTo) {
+        deleteBatch(from, maxTo, true)
+      } else {
+        val to = from + batchSize - 1
+        deleteBatch(from, to, false).flatMap(_ => deleteInBatches(to + 1, maxTo))
+      }
+    }
+
+    for {
+      toSeqNr <- highestSeqNrForDelete(persistenceId, toSequenceNr)
+      fromSeqNr <- lowestSequenceNrForDelete(persistenceId, toSeqNr, batchSize)
+      _ <- deleteInBatches(fromSeqNr, toSeqNr)
+    } yield ()
+  }
+
```

Comment:
I think we should only have the differences, which is mostly the sql statements, in the dialect. Share as much as possible/reasonable in a common base class. Since this is all internal api we are free to extract and refactor as needed when new dialects are added.

Reply:
My current plan is to keep the DAO interfaces like this, but have some abstract base classes with most of the logic and just provide capability to override specific parts that differ between the three supported dbs. I think that is little enough to be pretty clear. 

I also want to avoid OO-spaghetti calling back and forth a lot though, hard enough to understand the queries without tons of substitutions and parts that differ.